File: src/main/java/net/sf/prayer/Builtins.java
package net.sf.prayer;

import java.util.*;

public class Builtins {
    
    public Map<String, Operation> operations() {
        Map<String, Operation> res = new HashMap<>();
        
        res.put("print", new Operation() {
            public void execute(Context ctx) {
                System.out.print(ctx.getStack().pop());
            }
        });
        
        res.put("println", new Operation() {
            public void execute(Context ctx) {
                System.out.println(ctx.getStack().pop());
            }
        });
        
        res.put("+ add", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Object b = stack.pop();
                if (stack.peek() instanceof Integer && b instanceof Integer) {
                    stack.push(stack.popInt() + (Integer) b);
                } else {
                    stack.push(String.valueOf(stack.pop()) + String.valueOf(b));
                }
            }
        });
        
        res.put("- sub", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Integer b = stack.popInt();
                stack.push(stack.popInt() - b);
            }
        });
        
        res.put("* mul", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(stack.popInt() * stack.popInt());
            }
        });
        
        res.put("/ div", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Integer b = stack.popInt();
                stack.push(stack.popInt() / b);
            }
        });
        
        res.put("% mod", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Integer b = stack.popInt();
                stack.push(stack.popInt() % b);
            }
        });
        
        res.put("neg", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(- stack.popInt());
            }
        });
        
        res.put("= eq", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Object b = stack.pop();
                stack.push(b != null && b.equals(stack.pop()) ? 1 : 0);
            }
        });
        
        res.put("!= <> ne", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Object b = stack.pop();
                stack.push(b != null && !b.equals(stack.pop()) ? 1 : 0);
            }
        });
        
        res.put("! not", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(stack.popInt() == 0 ? 1 : 0);
            }
        });
        
        res.put("&& and", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                boolean a = stack.popInt() != 0;
                boolean b = stack.popInt() != 0;
                stack.push(a && b ? 1 : 0);
            }
        });
        
        res.put("|| or", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                boolean a = stack.popInt() != 0;
                boolean b = stack.popInt() != 0;
                stack.push(a || b ? 1 : 0);
            }
        });
        
        res.put("lt <", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Integer b = stack.popInt();
                stack.push(stack.popInt() < b ? 1 : 0);
            }
        });
        
        res.put("le <=", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Integer b = stack.popInt();
                stack.push(stack.popInt() <= b ? 1 : 0);
            }
        });
        
        res.put("gt >", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Integer b = stack.popInt();
                stack.push(stack.popInt() > b ? 1 : 0);
            }
        });
        
        res.put("ge >= =>", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Integer b = stack.popInt();
                stack.push(stack.popInt() >= b ? 1 : 0);
            }
        });
        
        res.put("itoa", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(Integer.toString(stack.popInt()));
            }
        });
        
        res.put("atoi", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(Integer.parseInt(stack.popStr()));
            }
        });
        
        res.put("isint ?int", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(stack.pop() instanceof Integer ? 1 : 0);
            }
        });
        
        res.put("isstr ?str", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(stack.pop() instanceof String ? 1 : 0);
            }
        });
        
        res.put("drop", new Operation() {
            public void execute(Context ctx) {
                ctx.getStack().pop();
            }
        });
        
        res.put("dup", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(stack.peek());
            }
        });
        
        res.put("over", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                stack.push(stack.get(stack.size() - 2));
            }
        });
        
        res.put("swap", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Object b = stack.pop(), a = stack.pop();
                stack.pushAll(b, a);
            }
        });
        
        res.put("rot", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                Object c = stack.pop(), b = stack.pop(), a = stack.pop();
                stack.pushAll(b, c, a);
            }
        });
        
        res.put("[] peek", new Operation() {
            public void execute(Context ctx) {
                OpStack stack = ctx.getStack();
                String idx = stack.pop().toString();
                stack.push(ctx.fetchVar(stack.pop().toString() + idx));
            }
        });
        
        res.put("#stacksize", new Operation() {
            public void execute(Context ctx) {
                System.err.println("STACK SIZE: " + ctx.getStack().size());
            }
        });
        
        res.put("#stackdata", new Operation() {
            public void execute(Context ctx) {
                System.err.println("STACK DATA: " + ctx.getStack());
            }
        });
        
        return res;
    }
    
}

File: javascript/builtins.js
function Builtins() {
}

Builtins.prototype.operations = function() {
    var res = [];
    var buf = '';

    res['print'] = {execute: function(ctx) {
        buf += ctx.stack.pop();
    }};

    res['println'] = {execute: function(ctx) {
        buf += ctx.stack.pop();
        alert(buf);
        buf = '';
    }};

    res['+ add'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop();
        stack.push(stack.pop() + b);
    }};

    res['- sub'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop();
        stack.push(stack.pop() - b);
    }};

    res['* mul'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(stack.pop() * stack.pop());
    }};

    res['/ div'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop();
        stack.push(Math.floor(stack.pop() / b));
    }};

    res['% mod'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop();
        stack.push(stack.pop() % b);
    }};
    
    res['neg'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(- stack.pop());
    }};
    
    res['= eq'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(stack.pop() === stack.pop() ? 1 : 0);
    }};
    
    res['!= <> ne'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(stack.pop() === stack.pop() ? 0 : 1);
    }};
    
    res['! not'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(!stack.pop());
    }};
    
    res['&& and'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var a = stack.pop() != 0;
        var b = stack.pop() != 0;
        stack.push(a && b ? 1 : 0);
    }};
    
    res['|| or'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var a = stack.pop() != 0;
        var b = stack.pop() != 0;
        stack.push(a || b ? 1 : 0);
    }};

    res['< lt'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop()
        stack.push(stack.pop() < b ? 1 : 0);
    }};
    
    res['<= le'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop()
        stack.push(stack.pop() <= b ? 1 : 0);
    }};
    
    res['> gt'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop()
        stack.push(stack.pop() > b ? 1 : 0);
    }};
    
    res['>= => ge'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop()
        stack.push(stack.pop() >= b ? 1 : 0);
    }};
    
    res['itoa'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push('' + stack.pop());
    }};
    
    res['atoi'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(parseInt(stack.pop()));
    }};
    
    res['?int isint'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(typeof(stack.pop()) === 'number' ? 1 : 0);
    }};
    
    res['?str isstr'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(typeof(stack.pop()) === 'string' ? 1 : 0);
    }};
    
    res['drop'] = {execute: function(ctx) {
        ctx.stack.pop();
    }};
    
    res['dup'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(stack[stack.length - 1]);
    }};

    res['over'] = {execute: function(ctx) {
        var stack = ctx.stack;
        stack.push(stack[stack.length - 2]);
    }};

    res['swap'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var b = stack.pop(), a = stack.pop();
        stack.push(b, a);
    }};

    res['rot'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var c = stack.pop(), b = stack.pop(), a = stack.pop();
        stack.push(b, c, a);
    }};
    
    res['[] peek'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var i = stack.pop(), a = stack.pop();
        stack.push(ctx.fetchVar(a + '' + i));
    }};

    res['#stacksize'] = {execute: function(ctx) {
        alert('STACK SIZE: ' + ctx.stack.length);
    }};
    
    res['#stackdata'] = {execute: function(ctx) {
        var stack = ctx.stack;
        var res = 'STACK DATA:';
        for (var i = 0; i < stack.length; i++) {
            res += '\n' + stack[i];
        }
        alert(res);
    }};
    
    return res;
}
